src/services/refactors/moveToNewFile.ts

```diff
@@ -1,11 +1,26 @@
 /* @internal */
 namespace ts.refactor {
     const refactorName = "Move to a new file";
+    const description = getLocaleSpecificMessage(Diagnostics.Move_to_a_new_file);
+
+    const moveToNewFileAction = {
+        name: refactorName,
+        description,
+        refactorKind: "refactor.move.newFile",
+    };
     registerRefactor(refactorName, {
+        refactorKinds: [moveToNewFileAction.refactorKind],
         getAvailableActions(context): readonly ApplicableRefactorInfo[] {
-            if (!context.preferences.allowTextChangesInNewFiles || getStatementsToMove(context) === undefined) return emptyArray;
-            const description = getLocaleSpecificMessage(Diagnostics.Move_to_a_new_file);
-            return [{ name: refactorName, description, actions: [{ name: refactorName, description }] }];
+            const statements = getStatementsToMove(context);
+            if (statements && context.preferences.allowTextChangesInNewFiles) {

```

Comment:
Checking the preference is much cheaper than computing the statements and should probably be checked first.